server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -352,24 +355,62 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
         }
     }
 
+    /**
+     * Merging the provided mappings. If the provided mapping list is ordered, merging will maintain the original ordering.
+     */
+    public DocumentMapper merge(String type, List<CompressedXContent> mappingSources, MergeReason reason) {
+        final DocumentMapper currentMapper = this.mapper;
+        if (currentMapper != null && mappingSources.size() == 1 && currentMapper.mappingSource().equals(mappingSources.get(0))) {
+            return currentMapper;
+        }
+
+        List<Map<String, Object>> convertedMappings = new LinkedList<>();
+        Explicit<Boolean> explicitSubobjects = null;
+        for (CompressedXContent mappingSource : mappingSources) {
+            Map<String, Object> mapping = MappingParser.convertToMap(mappingSource);
```

Comment:
__Note for reviewer:__ since `MappingParser.convertToMap()` is stateless, it should be safe to first apply it to the entire mappings list and only then apply parsing-and-merging for each of the mappings in the original order.
The former logic was applying convert-parse-merge for each of the mappings in order.

So far I only replaced one usage of the former logic (that relies on the existing `merge()` method) with this new logic (that relies on this `merge()` that accepts a list of mappings). There are additional cases where the older `merge()` override is applied sequentially on a list of mappings, where it may make sense to switch to this one as well. Please advice if and where else it makes sense to replace as well.

Reply:
I do agree with you, and looking at the code, whenever the previous `merge()` logic is applied sequentially on a list of mappings we can substitute it with the new logic. 
I am battled on whether this small refactor would find it's place in this PR. My suggestion here would be to create a new PR for this change and adding to it a test to make sure that calling sequentially the old `merge()` on a list of mappings has the same behaviour as calling the new `merge()` logic, to make sure that future changes to this code will stay consistent in both methods.

Reply:
I added a basic [test](https://github.com/elastic/elasticsearch/pull/97317/files#diff-904b4a9c1fa36a624cf1f095c37881f94fdc4530bb19d61ed5eb3be5216a7d6fR421-R429) to compare the two merge options and I opened a PR on this branch for the followup refactoring. Once we merge this PR, I'll convert the followup refactor PR to a real one.